tools/rewriter/DetermineAbi.cpp

```diff
@@ -261,5 +253,72 @@ FnSignature determineFnSignatureForProtoType(const clang::FunctionProtoType &fpt
       cgm,
       fpt.getCanonicalTypeUnqualified().castAs<clang::FunctionProtoType>());
 
-  return determineFnSignature(info, astContext, arch);
+  return determineAbiSignature(info, astContext, arch);
+}
+
+// Not thread safe.
+std::unordered_map<std::string, uint32_t> type_ids;
+
+uint32_t get_type_id(clang::QualType type) {
+  auto canonical_name = type.getCanonicalType().getAsString();
+  // Constructs the default value, 0, if the key doesn't exist yet.
+  auto type_id = type_ids[canonical_name];
+  if (type_id == 0) {
+    type_id = type_ids.size() + 1;
+    type_ids[canonical_name] = type_id;
+  }
+  return type_id;
```

Comment:
AFAIK this will collide for identically-named structs defined in different files. Canonical types are documented as being suitable for pointer-based comparison, so it might be reasonable to simply use `.getCanonicalType().getAsOpaquePtr()` here.

Reply:
Their lifetimes don't last long enough, though.

Reply:
That's what I had originally tried.

Reply:
That's surprising but unfortunate; in any case if we stick to hashing names it would be good to note with a TODO comment that there could be collisions.

Reply:
Sure, and I'll open an issue for this.

Reply:
Moved to #553.